src/awkward/_connect/numpy.py

```diff
@@ -141,9 +141,10 @@ def action(inputs, **ignore):
             return _array_ufunc_adjust(custom, inputs, kwargs, behavior)
 
         if ufunc is numpy.matmul:
-            custom_matmul = action_for_matmul(inputs)
-            if custom_matmul is not None:
-                return custom_matmul()
+            raise NotImplementedError
```

Comment:
I *think* we want to return `NotImplemented` rather than raise an error here
```suggestion
            return NotImplemented
```

The docs (https://numpy.org/doc/stable/user/basics.subclassing.html?highlight=numpy%20memmap%20all#array-ufunc-for-ufuncs, https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_ufunc__) aren't super clear on this, but I think that's the guidance. 

My understanding is that `return NotImplemented` will allow other arguments to implement `matmul`, e.g. an `ak.Array(...) @ np.array()` where the `ak.Array` is regular.

Reply:
It should raise an error because at the moment matmul is not implemented for any of the cases (not even the one you mention, ak.Array and np.array). I have always raised NotImplemented when working on awkward, but if the protocol has changed, I can change it as well. @jpivarski

Reply:
> My understanding is that `return NotImplemented` will allow other arguments to implement `matmul`, e.g. an `ak.Array(...) @ np.array()` where the `ak.Array` is regular.

It would exit the function on `return` as it does on `raise`

Reply:
N.b. the following is here for the benefit of keeping track of my thought processes :)

The `return NotImplemented` mechanism is usually used when there is the option for third-parties to implement the method, i.e. it's an agreed protocol:
```python
class IntLike:
    def __init__(self, value):
        self._value = value
    def __radd__(self, other):
        assert isinstance(other, int)
        return type(self)(self._value + other)
    def __repr__(self):
        return f"{type(self).__name__}({self._value!r})"

1 + IntLike(2)
```
i.e. `int` doesn't know about `IntLike`, so its `__add__` returns `NotImplemented` to allow `IntLike.__radd__(..., 1)` to be invoked. I *think* NumPy does something similar here; if we tell NumPy that we don't implement `matmul`, I would imagine that it will invoke NumPy's `matmul`. But, let me check!

I seemed to remember a NumPy passthrough that would see `np.asarray` being called on the Awkward types. This was the fallback mechanism that I expected to work here. However, my understanding of how this happens is not clear; if we return `NotImplemented` from the `__array_ufunc__()`, NumPy rightly doesn't know what to do with the Awkward Array type, and errors out. What I expected to see happen was that NumPy would try coercing each argument to a NumPy Array, and then try `np.matmul`. 

I'm fairly sure that there is a mechanism like this, but I think it might only be for `__array_function__`. Maybe, then, what we should do here is to try and coerce all `ak.Content`s into NumPy `ndarray`s, and call `np.matmul`? This would clearly need to be done outside the `apply`, as we need the entirety of each array to be regular.

Reply:
That's right, `return NotImplemented` lets other overloads take over, while `raise NotImplementedError` is a full-stop: no computation takes place. (These two cases unfortunately look very similar to one another.)

Doing a `return NotImplemented` would allow NumPy to take over, and it would try to cast the Awkward Array as a NumPy array by calling its `__array__` function. However, returning `NotImplemented` at this point in the code might not do it; we'd need to get the return value all the way back to the original `__array_ufunc__` call, if it needs to go through layers of the call stack in `broadcast_and_apply`, it might break something unrelated. To actually `return NotImplemented`, we might need to throw and catch an internal, custom exception, just so that we can skip most of the call stack and get the `NotImplemented` object directly out of `__array_ufunc__`.

Asking the question, "What _should_ we do?" I'm leaning more toward `raise NotImplementedError`, so that `@` always fails with an error message that indicates that this operator can and will be handled in the future; we might end up soliciting requests to implement it sooner (finding out just how many people need it). Doing `return NotImplemented` (properly) would make `@` work for regular arrays and not work for irregular arrays with a weird error message, most likely, "This array is not convertible to NumPy."

But in order for that to be useful, this PR should add an error message:

```python
            raise ak._errors.wrap_error(
                NotImplementedError(
                    "matrix multiplication (`@` or `np.matmul`) is not yet implemented for Awkward Arrays"
                )
            )
```

> I'm fairly sure that there is a mechanism like this, but I think it might only be for `__array_function__`.

Yes, the mechanism that returns `NotImplemented` to allow NumPy to try to cast the Awkward Array and use its own function is for NEP 18 (`__array_function__`), not NEP 13 (`__array_ufunc__`). With the exception of matrix multiplication, all ufuncs go through a common mechanism that applies to all Awkward Arrays except those with records (because we deliberately prevented it). None of the ufuncs (currently) `return NotImplemented` to let NumPy cast it.

Reply:
> None of the ufuncs (currently) `return NotImplemented` to let NumPy cast it.

This was my motivation when choosing `raise`.

Reply:
@jpivarski I tried enabling `return NotImplemented`, but as we both remark, it looks like ufuncs don't implement the `np.asarray` fallback that would catch this case, so it fails just the same.
> Doing return NotImplemented (properly) would make @ work for regular arrays and not work for irregular arrays with a weird error message, most likely, "This array is not convertible to NumPy."

We could add an exception for the non-NumPy case, which is what we do for `__array_function__` cases.

@ioanaif thanks for the discussion; the question of whether to do fallbacks is a bit murky, and I can see that there are varying opinions on this one.